### PR TITLE
[AMDGPU] Make AMDGPULowerKernelArguments a module pass

### PR DESCRIPTION
**Pull request by kerbowa**
After c4d89203f3 AMDGPULowerKernelArguments may clone functions and modify the kernel signature of those functions to enable preloading hidden kernel arguments. These leftover functions end up as dead declarations which may cause issues with the toolchain downstream.

This patch makes AMDGPULowerKernelArguments a module pass so that we can safely erase these leftover declarations.

There is also some small refactoring to avoid duplicated logic with the different pass managers. The update changes the pass interfaces to look similar to other AMDGPU passes that have been migrated over to the new pass manager.

### CODE REVIEW

**Comment by shiltian** [comment]
I'd make the refactoring and change to module pass separate PRs.

**Comment by kerbowa** [comment]
> I'd make the refactoring and change to module pass separate PRs.

I can, but the refactoring is actually quite minor, the diff looks big but 90% is just indentation.

File: llvm/test/CodeGen/AMDGPU/llc-pipeline.ll [hunk 1]
```diff
@@ -1170,8 +1166,7 @@
 ; GCN-O3-NEXT:    CallGraph Construction
 ; GCN-O3-NEXT:    Call Graph SCC Pass Manager
 ; GCN-O3-NEXT:      AMDGPU Annotate Kernel Features
-; GCN-O3-NEXT:      FunctionPass Manager
-; GCN-O3-NEXT:        AMDGPU Lower Kernel Arguments
+; GCN-O3-NEXT:    AMDGPU Lower Kernel Arguments
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/llc-pipeline.ll:1169]: Can this be a CGSCC pass instead?

File: llvm/test/CodeGen/AMDGPU/preload-implicit-kernargs-IR-lowering.ll [hunk 2]
```diff
@@ -83,6 +88,8 @@ define amdgpu_kernel void @preloadremainder_z(ptr addrspace(1) %out) {
   ret void
 }
 
+; PRELOAD-NOT: declare {{.*}}@1
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/preload-implicit-kernargs-IR-lowering.ll:91]: better to use -implicit-check-not=declare

Comment (kerbowa) [reply llvm/test/CodeGen/AMDGPU/preload-implicit-kernargs-IR-lowering.ll:91]: This doesn't work because of implicit_arg ptr declaration.

File: llvm/test/CodeGen/AMDGPU/llc-pipeline.ll (hunk 1 above)
Comment (kerbowa) [reply llvm/test/CodeGen/AMDGPU/llc-pipeline.ll]: I've made this change. I'm worried that this leaves the SCC in an inconsonant state, although it works. It breaks if I try to remove any functions outside of finalization.

There are 3 other options. Was wondering if you have any opinion on the best one.
1. Module pass
2. Keep it a function pass and do all the preload stuff in the doInitialization callback
3. Make the preloading stuff a separate module pass and keep lowerkernelarguments a function pass
4. CGSCC pass

I'm leaning towards option 3 personally.

Comment (JanekvO) [reply llvm/test/CodeGen/AMDGPU/llc-pipeline.ll]: wrt some of the (legacy) passes: it may not be an issue for now I'd be cautious adding module passes in the (Pre)CodeGen part as it pops off the CGSCC pass off the pass manager stack. From what I see in `llc-pipeline.ll`, it shouldn't be an issue but it's the reason why a few lines under where you add the pass in `AMDGPUTargetMachine` you see a sudden `addPass(new DummyCGSCCPass());` added: it pushes the CGSCC pass back on the pass manager stack.

Related to that: having the pass be a function pass will more than likely run the functions in CGSCC order anyways since it seems to inherit the CGSCC pass manager from the `AMDGPUAnnotateKernelFeatures` pass before it.

TL;DR: fine to add module pass in this case but may need a comment in `AMDGPUTargetMachine` mentioning that it pops off the CGSCC pass manager from the PM stack for future explorers of said code.

Comment (kerbowa) [reply llvm/test/CodeGen/AMDGPU/llc-pipeline.ll]: Thanks @JanekvO I'll add a comment to that effect.

@arsenm to future proof for NPM I think our only options are making AMDGPULowerKernelArguments a module pass or factoring out the preload stuff and creating a new module pass. What do you think is better?

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/llc-pipeline.ll]: I thought you were moving the preload logic to the attributor? This probably belongs in the attributor, or a separate CGSCC/CGSCC-in-module pass.

This does not feel like it belongs in AMDGPULowerKernelArguments. AMDGPULowerKernelArguments is at least a 50% hack due to clang emitting sub-optimal kernel signatures to begin with (which #115821 will soon give us a path forward to fixing). The other piece is it enables vectorization of kernel argument loads, which will not go away. In either case, its only job is to rewrite argument references in terms of a base pointer. It shouldn't need to consider the use context like preload should

File: llvm/test/CodeGen/AMDGPU/preload-implicit-kernargs-IR-lowering.ll (hunk 2 above)
Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/preload-implicit-kernargs-IR-lowering.ll:91]: Can you relax it to `@{{[0-9]+}}`?

**Comment by krzysz00** [comment]
One note - though it's late - `CallGraphUpdater` exists
